panther-labs/panther
# Add custom certificate CFN resource

## Background

Use a CloudFormation custom resource for the default self-signed certificate to enable a self-contained deployment model.

Closes: #847 

This is a much more natural model in any case - our current way of handling certificates requires adding them as a parameter and an output to the same stack and then listing all ACM/IAM certificates when manually tearing down. Now CloudFormation neatly tracks which cert is actually used.

## Changes

- Add a `panther-cfn-custom-resources` Lambda function which supports the `Custom::Certificate` resource type (more coming soon). This function is created in the `bootstrap-gateway` stack (it can't be in the first stack because an S3 bucket needs to exist first so `mage deploy` can package it when deploying from source)
    - Moved certificate create/destroy out of `mage` and into the Lambda function
- Moved the load balancer listener (the thing which actually uses the cert) from `bootstrap` to the `web` stack because now it depends on the custom-resource
- Bugfix: Create the correct type of changeset after a bad stack has been automatically deleted by `mage deploy` (I know @nhakmiller ran into this issue recently)
- Upgrade our `backoff` to v4 - we were still on v2
- Fix deploy/teardown progress indicators to be consistent

## Testing

The resource shows up in CloudFormation like this:

<img width="1191" alt="Screen Shot 2020-05-04 at 5 56 09 PM" src="https://user-images.githubusercontent.com/3608925/81352649-e96f8b00-907b-11ea-8a22-29114c97c158.png">

- Deploy with cert specified in the config file
- Deploy + teardown with generated IAM certs
- Deploy + teardown with generated ACM certs
- Generated errors when creating and deleting certs to see how they look in CloudFormation:

<img width="1323" alt="Screen Shot 2020-05-04 at 5 56 01 PM" src="https://user-images.githubusercontent.com/3608925/81352570-c04efa80-907b-11ea-91c6-7526fa984603.png">

### Migration
- Deployed this branch on top of a `master` deployment to make sure the migration is smooth

The web URL will not change, but it will be unreachable during the deploy as the listener moves to a different stack. If someone specifies a certificate in the settings file, it is used exactly the same as before.

If a user was relying on the default self-signed certificate, that cert will no longer be used - a new one will be generated instead. They will need to manually delete their old cert if they care to do so.

We could "import" their existing self-signed cert if we want by changing the "CREATE" functionality of the Custom::Certificate resource to scan for existing certs before generating a new one. I chose not to do this for a few reasons:

- Production deployments shouldn't be relying on self-signed certs anyway, no need to have lots of code for backwards-compatibility with the non-secure default
- Here we have a guarantee that the private key is completely secure - it's generated in Lambda memory (never stored on disk), then sent to ACM/IAM and never seen again. This is more secure for default deployments because now the key can't be stolen from a developer laptop.
